runtime/vm/createramclass.cpp

```diff
@@ -1614,6 +1614,44 @@ isClassPermittedBySealedSuper(J9ROMClass *superRomClass, U_8* className, U_16 cl
 	return result;
 }
 
+#if JAVA_SPEC_VERSION >= 16
+/**
+ * JEP 397 (second preview): if super class/interface is sealed, IncompatibleClassChangeError is throw out
+ * if one of the following conditions is false:
+ * (1) the inheriting subclass is not in the same module as its super class/interface;
+ * (2) the inheriting subclass (non-public) is not in the same package as its super class/interface.
+ *
+ * @param vmThread the current VM thread
+ * @param superClass the super class/interface
+ * @param romClass the ROM class of subclass
+ * @param module the subclass's module
+ * @param packageID the subclass's package ID
+ * @return TRUE if subclass can legally inherit the super, FALSE otherwise.
+ */
+static VMINLINE BOOLEAN
+isClassInTheSameModuleOrPckageAsSealedSuper(J9VMThread *vmThread, J9Class *superClass, J9ROMClass *romClass, J9Module *module, UDATA packageID)
+{
+	if (J9ROMCLASS_IS_SEALED(superClass->romClass)) {
+		J9UTF8 *className = J9ROMCLASS_CLASSNAME(romClass);
+		bool classIsPublic = J9_ARE_ALL_BITS_SET(romClass->modifiers, J9AccPublic);
+
+		if (module != superClass->module) {
```

Comment:
My understanding was that sealed classes cared about being in the same module (then any relationship was ok) and if they weren't in the same module, they had to be in the same package.

That's not what this check says - can you confirm this is correct?  Is there a spec reference for this?

And I would also expect a isNamed check on the module...

Reply:
As mentioned at https://github.com/eclipse/openj9/issues/11273#issuecomment-734468800 as to http://cr.openjdk.java.net/~gbierman/jep397/jep397-20201204/specs/sealed-classes-jvms.html, the new changes for sealed classes from VM perspective are as follows:
```
5.3.5 Deriving a Class from a class File Representation
...
3. If C has a direct superclass, the symbolic reference from C to its direct superclass is resolved ...
...
    Otherwise, if the class named as the direct superclass of C has a PermittedSubclasses attribute (4.7.30)
    and any of the following are true, derivation fails with a IncompatibleClassChangeError:
---> (1)The superclass belongs to a different run-time module than C.
---> (2)C does not have its ACC_PUBLIC flag set (4.1) and the superclass belongs to a different run-time package than C.
...
4. If C has any direct superinterfaces, the symbolic references from C to its direct superinterfaces are resolved
...
  Otherwise, for each direct superinterface, if the superinterface has a PermittedSubclasses attribute (4.7.30) 
   and any of the following are true, derivation fails with a IncompatibleClassChangeError:
---> (1) The superinterface belongs to a different run-time module than C.
---> (2) C does not have its ACC_PUBLIC flag set (4.1) and the superinterface belongs to a different run-time package than C.
...
```

Reply:
The description in there might be ambiguous but it seems there is no dependency between condition (1) and (2).

Reply:
`isModuleNamed()` is added to check whether the superclass module is named or not.

Reply:
I think we can add one more test case to check whether it works in different package is their modules are identical in the case of the superclass's named module against the RI's behavior.

Reply:
Hi @DanHeidinga,  the code has been updated to check the named module with `J9_IS_J9MODULE_UNNAMED
` and a new test  `testSubClassOfSealedSuperFromDifferentPackageInSameNamedModule` is added to check the changes (passed)
[new_tests_jdk16_hotspot.txt](https://github.com/eclipse/openj9/files/6087490/new_tests_jdk16_hotspot.txt)
[new_tests_jdk16_openj9.txt](https://github.com/eclipse/openj9/files/6087491/new_tests_jdk16_openj9.txt)